File: win/UI/ChmodWindow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using PixShell.Logging;
using PixShell.Sftp;

namespace PixShell.UI;

/// <summary>
/// SFTP「文件权限…」弹窗（对齐 mac UI/ChmodWindow.swift）。
/// 纯代码构建：ToolWindow + CanResizeWithGrip；Owner/Group/Other rwx 复选 + 实时八进制；
/// 递归 + 应用范围；OK 经 SSH ExecAsync 跑 chmod / chmod -R / find -type f|d -exec chmod。
/// 铁律：所有色值走 DynamicResource（SetResourceReference），零色差。
/// </summary>
public sealed class ChmodWindow
{
    private Window? _window;
    private TextBlock? _pathLabel;
    private TextBlock? _octalLabel;
    private CheckBox[] _ownerBoxes = Array.Empty<CheckBox>();
    private CheckBox[] _groupBoxes = Array.Empty<CheckBox>();
    private CheckBox[] _otherBoxes = Array.Empty<CheckBox>();
    private CheckBox? _recursiveBox;
    private RadioButton? _scopeBoth;
    private RadioButton? _scopeFiles;
    private RadioButton? _scopeDirs;
    private StackPanel? _scopePanel;

    private List<string> _paths = new();
    private uint _initialMode = 0x1ED; // 0755
    private Func<string, Task<string>>? _execRunner;
    private Action<string>? _onDone;
    private bool _busy;

    /// <summary>展示：paths 为远端绝对路径；mode 为 POSIX 权限位（可含文件类型高位，取低 9 位）。</summary>
    public void Show(
        Window? owner,
        IReadOnlyList<string> paths,
        uint mode,
        Func<string, Task<string>>? execRunner,
        Action<string>? onDone = null)
    {
        if (paths == null || paths.Count == 0) return;
        _paths = paths.ToList();
        _execRunner = execRunner;
        _onDone = onDone;
        _initialMode = mode & 0x1FF; // 0777
        if (_window == null) Build(owner);
        else if (owner != null && !ReferenceEquals(_window.Owner, owner)) _window.Owner = owner;

        ApplyMode(_initialMode);
        var shown = string.Join("\n", _paths.Take(3));
        if (_paths.Count > 3) shown += $"\n…共 {_paths.Count} 项";
        _pathLabel!.Text = shown;
        _recursiveBox!.IsChecked = false;
        _scopeBoth!.IsChecked = true;
        UpdateScopeEnabled();
        _busy = false;

        _window!.Show();
        _window.Activate();
    }

    public void Hide() => _window?.Hide();

    private void Build(Window? owner)
    {
        var title = new TextBlock
        {
            Text = "修改文件权限",
            FontSize = 15,
            FontWeight = FontWeights.SemiBold,
            VerticalAlignment = VerticalAlignment.Center,
        };
        BindText(title);

        var closeBtn = new Button
        {
            Content = "取消",
            Style = (Style)Application.Current.Resources["PillButton"],
            Margin = new Thickness(0, 0, 0, 0),
            IsCancel = true,
        };
        closeBtn.Click += (_, _) => Hide();

        var head = new DockPanel { Margin = new Thickness(0, 0, 0, 10) };
        DockPanel.SetDock(closeBtn, Dock.Right);
        head.Children.Add(closeBtn);
        head.Children.Add(title);

        _pathLabel = new TextBlock
        {
            FontSize = 11,
            FontFamily = (FontFamily)Application.Current.Resources["FontMono"],
            TextWrapping = TextWrapping.Wrap,
            Margin = new Thickness(0, 0, 0, 8),
            MaxHeight = 64,
        };
        BindMuted(_pathLabel);

        _ownerBoxes = MakeTriad();
        _groupBoxes = MakeTriad();
        _otherBoxes = MakeTriad();

        var octTitle = new TextBlock
        {
            Text = "权限值",
            FontSize = 11,
            FontWeight = FontWeights.SemiBold,
            VerticalAlignment = VerticalAlignment.Center,
        };
        BindText(octTitle);
        _octalLabel = new TextBlock
        {
            Text = "0755",
            FontSize = 18,
            FontFamily = (FontFamily)Application.Current.Resources["FontMono"],
            HorizontalAlignment = HorizontalAlignment.Right,
            VerticalAlignment = VerticalAlignment.Center,
        };
        _octalLabel.SetResourceReference(TextBlock.ForegroundProperty, "BrushAccent");

        var octRow = new DockPanel { Margin = new Thickness(0, 4, 0, 8) };
        DockPanel.SetDock(_octalLabel, Dock.Right);
        octRow.Children.Add(_octalLabel);
        octRow.Children.Add(octTitle);

        _recursiveBox = new CheckBox
        {
            Content = "递归设置子目录",
            FontSize = 12,
            Margin = new Thickness(0, 2, 0, 6),
            VerticalContentAlignment = VerticalAlignment.Center,
        };
        BindText(_recursiveBox);
        _recursiveBox.Checked += (_, _) => UpdateScopeEnabled();
        _recursiveBox.Unchecked += (_, _) => UpdateScopeEnabled();

        _scopeBoth = MakeScopeRadio("应用于文件和目录", true);
        _scopeFiles = MakeScopeRadio("只应用于文件", false);
        _scopeDirs = MakeScopeRadio("只应用于目录", false);
        _scopePanel = new StackPanel { Margin = new Thickness(18, 0, 0, 8) };
        _scopePanel.Children.Add(_scopeBoth);
        _scopePanel.Children.Add(_scopeFiles);
        _scopePanel.Children.Add(_scopeDirs);

        var okBtn = new Button
        {
            Content = "确定",
            Style = (Style)Application.Current.Resources["PillButton"],
            Tag = "Primary",
            MinWidth = 72,
            IsDefault = true,
            Margin = new Thickness(8, 0, 0, 0),
        };
        okBtn.Click += async (_, _) => await ConfirmAsync();
        var cancelBtn = new Button
        {
            Content = "取消",
            Style = (Style)Application.Current.Resources["PillButton"],
            MinWidth = 64,
            IsCancel = true,
        };
        cancelBtn.Click += (_, _) => Hide();

        var foot = new StackPanel
        {
            Orientation = Orientation.Horizontal,
            HorizontalAlignment = HorizontalAlignment.Right,
            Margin = new Thickness(0, 8, 0, 0),
        };
        foot.Children.Add(cancelBtn);
        foot.Children.Add(okBtn);

        var body = new StackPanel();
        body.Children.Add(head);
        body.Children.Add(_pathLabel);
        body.Children.Add(TriadRow("Owner", _ownerBoxes));
        body.Children.Add(TriadRow("Group", _groupBoxes));
        body.Children.Add(TriadRow("Other", _otherBoxes));
        body.Children.Add(octRow);
        body.Children.Add(_recursiveBox);
        body.Children.Add(_scopePanel);
        body.Children.Add(foot);

        var scroll = new ScrollViewer
        {
            VerticalScrollBarVisibility = ScrollBarVisibility.Auto,
            Content = new Border
            {
                Padding = new Thickness(14),
                Child = body,
            },
        };

        _window = new Window
        {
            Title = "修改文件权限",
            Owner = owner,
            Width = 360,
            Height = 420,
            MinWidth = 340,
            MinHeight = 380,
            WindowStartupLocation = owner != null
                ? WindowStartupLocation.CenterOwner
                : WindowStartupLocation.CenterScreen,
            WindowStyle = WindowStyle.ToolWindow,
            ResizeMode = ResizeMode.CanResizeWithGrip,
            ShowInTaskbar = false,
            Content = scroll,
        };
        _window.SetResourceReference(Control.BackgroundProperty, "BrushBg");
        // 关掉只隐藏，下次复用
        _window.Closing += (_, e) => { e.Cancel = true; _window!.Hide(); };
    }

    private CheckBox[] MakeTriad()
    {
        // bit tags: r=4 w=2 x=1
        var labels = new[] { ("读 (r=4)", 4), ("写 (w=2)", 2), ("执行 (x=1)", 1) };
        var boxes = new CheckBox[3];
        for (var i = 0; i < 3; i++)
        {
            var (text, bit) = labels[i];
            var box = new CheckBox
            {
                Content = text,
                Tag = bit,
                FontSize = 12,
                Margin = new Thickness(0, 0, 10, 0),
                VerticalContentAlignment = VerticalAlignment.Center,
            };
            BindText(box);
            box.Checked += (_, _) => RefreshOctal();
            box.Unchecked += (_, _) => RefreshOctal();
            boxes[i] = box;
        }
        return boxes;
    }

    private static UIElement TriadRow(string name, CheckBox[] boxes)
    {
        var lab = new TextBlock
        {
            Text = name,
            FontSize = 11,
            FontWeight = FontWeights.SemiBold,
            Width = 52,
            VerticalAlignment = VerticalAlignment.Center,
            Margin = new Thickness(0, 0, 6, 0),
        };
        BindText(lab);
        var row = new StackPanel
        {
            Orientation = Orientation.Horizontal,
            Margin = new Thickness(0, 0, 0, 6),
        };
        row.Children.Add(lab);
        foreach (var b in boxes) row.Children.Add(b);
        return row;
    }

    private RadioButton MakeScopeRadio(string text, bool isChecked)
    {
        var rb = new RadioButton
        {
            Content = text,
            GroupName = "ChmodScope",
            FontSize = 12,
            IsChecked = isChecked,
            Margin = new Thickness(0, 0, 0, 4),
            VerticalContentAlignment = VerticalAlignment.Center,
        };
        BindText(rb);
        return rb;
    }

    private void UpdateScopeEnabled()
    {
        var on = _recursiveBox?.IsChecked == true;
        if (_scopePanel != null) _scopePanel.IsEnabled = on;
    }

    private uint CurrentMode()
    {
        static uint Sum(CheckBox[] boxes)
        {
            uint n = 0;
            foreach (var b in boxes)
                if (b.IsChecked == true && b.Tag is int bit) n += (uint)bit;
            return n;
        }
        return (Sum(_ownerBoxes) << 6) | (Sum(_groupBoxes) << 3) | Sum(_otherBoxes);
    }

    private void ApplyMode(uint mode)
    {
        var m = mode & 0x1FF; // 0777
        static void Set(CheckBox[] boxes, uint nibble)
        {
            foreach (var b in boxes)
            {
                var bit = b.Tag is int t ? (uint)t : 0;
                b.IsChecked = (nibble & bit) != 0;
            }
        }
        Set(_ownerBoxes, (m >> 6) & 0x7);
        Set(_groupBoxes, (m >> 3) & 0x7);
        Set(_otherBoxes, m & 0x7);
        RefreshOctal();
    }

    private void RefreshOctal()
    {
        if (_octalLabel == null) return;
        _octalLabel.Text = $"0{Convert.ToString(CurrentMode(), 8).PadLeft(3, '0')}";
    }

    private async Task ConfirmAsync()
    {
        if (_busy) return;
        if (_paths.Count == 0) { Hide(); return; }

        var mode = CurrentMode();
        var oct = Convert.ToString(mode, 8).PadLeft(3, '0');
        var recursive = _recursiveBox?.IsChecked == true;
        var scope = 0; // 0 both / 1 files / 2 dirs
        if (_scopeFiles?.IsChecked == true) scope = 1;
        else if (_scopeDirs?.IsChecked == true) scope = 2;

        var quoted = string.Join(" ", _paths.Select(SftpTransfer.Quote));
        string cmd;
        if (recursive)
        {
            cmd = scope switch
            {
                1 => $"find {quoted} -type f -exec chmod {oct} {{}} + 2>&1",
                2 => $"find {quoted} -type d -exec chmod {oct} {{}} + 2>&1",
                _ => $"chmod -R {oct} {quoted} 2>&1",
            };
        }
        else
        {
            cmd = $"chmod {oct} {quoted} 2>&1";
        }

        if (_execRunner == null)
        {
            _onDone?.Invoke("需要 SSH 会话才能改权限");
            Hide();
            return;
        }

        _busy = true;
        Log.Info($"chmod {oct}{(recursive ? " -R" : "")} {_paths.Count} 项", "sftp");
        try
        {
            var outp = await _execRunner(cmd).ConfigureAwait(true);
            var trimmed = (outp ?? "").Trim();
            var msg = string.IsNullOrEmpty(trimmed)
                ? $"已设置权限 0{oct}"
                : "chmod: " + trimmed;
            _onDone?.Invoke(msg);
            Hide();
        }
        catch (Exception ex)
        {
            Log.Warn($"chmod 失败: {ex.Message}", "sftp");
            _onDone?.Invoke("chmod 失败: " + ex.Message);
            Hide();
        }
        finally
        {
            _busy = false;
        }
    }

    private static void BindText(FrameworkElement el)
    {
        if (el is Control c)
            c.SetResourceReference(Control.ForegroundProperty, "BrushText");
        else if (el is TextBlock tb)
            tb.SetResourceReference(TextBlock.ForegroundProperty, "BrushText");
    }

    private static void BindMuted(TextBlock tb)
        => tb.SetResourceReference(TextBlock.ForegroundProperty, "BrushMuted");
}

File: mac/Sources/PixShell/UI/ChmodWindow.swift
import AppKit

/// SFTP「文件权限…」弹窗。
/// ~340×380，可拖、可缩放；Theme.bg / Theme.text，无杂色。
/// Owner/Group/Other 各 r/w/x（4/2/1）复选 + 实时八进制；递归 + 应用范围；OK 后走 chmod / chmod -R。
final class ChmodWindow: NSWindowController {

    private let card = NSView()
    private let pathLabel = NSTextField(wrappingLabelWithString: "")
    private let octalLabel = NSTextField(labelWithString: "0755")
    private var ownerBoxes: [NSButton] = []
    private var groupBoxes: [NSButton] = []
    private var otherBoxes: [NSButton] = []
    private let recursiveBox = NSButton(checkboxWithTitle: "递归设置子目录", target: nil, action: nil)
    private let modePopup = NSPopUpButton()

    private var paths: [String] = []
    private var initialMode: UInt32 = 0o755
    /// 执行 chmod 命令（宿主注入 ssh.exec）
    var execRunner: ((String, @escaping (String) -> Void) -> Void)?
    var onDone: ((String) -> Void)?

    init() {
        let w = NSWindow(contentRect: NSRect(x: 0, y: 0, width: 340, height: 380),
                         styleMask: [.titled, .closable, .fullSizeContentView, .resizable],
                         backing: .buffered, defer: false)
        w.title = "文件权限"
        w.titlebarAppearsTransparent = true
        w.titleVisibility = .hidden
        w.isMovableByWindowBackground = true
        w.backgroundColor = .clear
        w.isOpaque = false
        w.hasShadow = true
        w.isReleasedWhenClosed = false
        w.minSize = NSSize(width: 300, height: 320)
        w.standardWindowButton(.closeButton)?.isHidden = true
        w.standardWindowButton(.miniaturizeButton)?.isHidden = true
        w.standardWindowButton(.zoomButton)?.isHidden = true
        super.init(window: w)
        build()
    }
    required init?(coder: NSCoder) { fatalError() }

    /// 展示：paths 为远端绝对路径；mode 为 POSIX 权限位（可含文件类型高位，取低 9 位）。
    func show(paths: [String], mode: UInt32, execRunner: ((String, @escaping (String) -> Void) -> Void)?) {
        self.paths = paths
        self.execRunner = execRunner
        self.initialMode = mode & 0o777
        applyMode(initialMode)
        let shown = paths.prefix(3).joined(separator: "\n")
        let more = paths.count > 3 ? "\n…共 \(paths.count) 项" : ""
        pathLabel.stringValue = shown + more
        recursiveBox.state = .off
        modePopup.selectItem(at: 0)
        window?.center()
        window?.makeKeyAndOrderFront(nil)
        NSApp.activate(ignoringOtherApps: true)
    }

    func hide() { window?.orderOut(nil) }

    private func build() {
        guard let w = window else { return }
        card.rounded(Theme.radiusLg, bg: Theme.bg, border: Theme.borderStrong)
        card.translatesAutoresizingMaskIntoConstraints = false

        let root = EscapableView()
        root.onEscape = { [weak self] in self?.cancel() }
        root.addSubview(card)
        NSLayoutConstraint.activate([
            card.topAnchor.constraint(equalTo: root.topAnchor),
            card.bottomAnchor.constraint(equalTo: root.bottomAnchor),
            card.leadingAnchor.constraint(equalTo: root.leadingAnchor),
            card.trailingAnchor.constraint(equalTo: root.trailingAnchor),
        ])
        w.contentView = root

        let title = NSTextField(labelWithString: "文件权限…")
        title.font = Theme.ui(15, .semibold); title.textColor = Theme.text
        let closeBtn = PillButton("取消", style: .secondary, hPad: 10, target: self, action: #selector(cancel))
        let head = NSStackView(views: [title, NSView(), closeBtn]); head.spacing = 12; head.alignment = .centerY
        head.translatesAutoresizingMaskIntoConstraints = false

        pathLabel.font = Theme.mono(11); pathLabel.textColor = Theme.muted
        pathLabel.translatesAutoresizingMaskIntoConstraints = false
        pathLabel.maximumNumberOfLines = 4

        ownerBoxes = makeTriad(title: "Owner（所有者）", bits: [4, 2, 1])
        groupBoxes = makeTriad(title: "Group（组）", bits: [4, 2, 1])
        otherBoxes = makeTriad(title: "Other（其他）", bits: [4, 2, 1])
        let ownerRow = triadRow("Owner", boxes: ownerBoxes)
        let groupRow = triadRow("Group", boxes: groupBoxes)
        let otherRow = triadRow("Other", boxes: otherBoxes)

        let octTitle = NSTextField(labelWithString: "权限值")
        octTitle.font = Theme.ui(11, .semibold); octTitle.textColor = Theme.text
        octalLabel.font = Theme.mono(18); octalLabel.textColor = Theme.accent
        octalLabel.alignment = .center
        let octRow = NSStackView(views: [octTitle, NSView(), octalLabel]); octRow.alignment = .centerY
        octRow.translatesAutoresizingMaskIntoConstraints = false

        recursiveBox.font = Theme.ui(12); recursiveBox.contentTintColor = Theme.text
        styleCheck(recursiveBox)
        recursiveBox.target = self; recursiveBox.action = #selector(recursiveToggled)

        modePopup.removeAllItems()
        modePopup.addItems(withTitles: [
            "应用到文件和目录",
            "只应用到文件",
            "只应用到目录",
        ])
        modePopup.font = Theme.ui(12)
        // 强制走 Theme，避免系统 control 与卡片底色错配
        modePopup.appearance = NSAppearance(named: Theme.dark ? .darkAqua : .aqua)
        modePopup.isEnabled = false
        modePopup.translatesAutoresizingMaskIntoConstraints = false
        modePopup.widthAnchor.constraint(greaterThanOrEqualToConstant: 180).isActive = true

        let okBtn = PillButton("确定", style: .primary, hPad: 16, target: self, action: #selector(confirm))
        let cancelBtn = PillButton("取消", style: .secondary, hPad: 14, target: self, action: #selector(cancel))
        let foot = NSStackView(views: [NSView(), cancelBtn, okBtn]); foot.spacing = 8; foot.alignment = .centerY
        foot.translatesAutoresizingMaskIntoConstraints = false

        let body = NSStackView(views: [
            head, pathLabel, ownerRow, groupRow, otherRow, octRow,
            recursiveBox, modePopup, foot,
        ])
        body.orientation = .vertical; body.alignment = .leading; body.spacing = 12
        body.translatesAutoresizingMaskIntoConstraints = false
        card.addSubview(body)
        NSLayoutConstraint.activate([
            body.topAnchor.constraint(equalTo: card.topAnchor, constant: 14),
            body.leadingAnchor.constraint(equalTo: card.leadingAnchor, constant: 14),
            body.trailingAnchor.constraint(equalTo: card.trailingAnchor, constant: -14),
            body.bottomAnchor.constraint(equalTo: card.bottomAnchor, constant: -14),
            head.widthAnchor.constraint(equalTo: body.widthAnchor),
            pathLabel.widthAnchor.constraint(equalTo: body.widthAnchor),
            ownerRow.widthAnchor.constraint(equalTo: body.widthAnchor),
            groupRow.widthAnchor.constraint(equalTo: body.widthAnchor),
            otherRow.widthAnchor.constraint(equalTo: body.widthAnchor),
            octRow.widthAnchor.constraint(equalTo: body.widthAnchor),
            modePopup.widthAnchor.constraint(equalTo: body.widthAnchor),
            foot.widthAnchor.constraint(equalTo: body.widthAnchor),
        ])
    }

    private func makeTriad(title: String, bits: [Int]) -> [NSButton] {
        bits.map { bit in
            let label: String
            switch bit {
            case 4: label = "读 (r=4)"
            case 2: label = "写 (w=2)"
            default: label = "执行 (x=1)"
            }
            let b = NSButton(checkboxWithTitle: label, target: self, action: #selector(bitChanged))
            b.tag = bit
            styleCheck(b)
            return b
        }
    }

    private func styleCheck(_ b: NSButton) {
        b.font = Theme.ui(12)
        b.contentTintColor = Theme.text
        // 强制标题色与 Theme 一致，避免系统控件浅色/深色错配
        b.attributedTitle = NSAttributedString(string: b.title, attributes: [
            .foregroundColor: Theme.text,
            .font: Theme.ui(12),
        ])
    }

    private func triadRow(_ name: String, boxes: [NSButton]) -> NSView {
        let lab = NSTextField(labelWithString: name)
        lab.font = Theme.ui(11, .semibold); lab.textColor = Theme.text
        lab.translatesAutoresizingMaskIntoConstraints = false
        lab.widthAnchor.constraint(equalToConstant: 52).isActive = true
        let row = NSStackView(views: [lab] + boxes)
        row.spacing = 10; row.alignment = .centerY
        row.translatesAutoresizingMaskIntoConstraints = false
        return row
    }

    @objc private func bitChanged() { refreshOctal() }
    @objc private func recursiveToggled() {
        modePopup.isEnabled = recursiveBox.state == .on
    }

    private func currentMode() -> UInt32 {
        func sum(_ boxes: [NSButton]) -> UInt32 {
            boxes.reduce(0) { $0 + ($1.state == .on ? UInt32($1.tag) : 0) }
        }
        return (sum(ownerBoxes) << 6) | (sum(groupBoxes) << 3) | sum(otherBoxes)
    }

    private func applyMode(_ mode: UInt32) {
        let m = mode & 0o777
        func set(_ boxes: [NSButton], _ nibble: UInt32) {
            for b in boxes { b.state = (nibble & UInt32(b.tag)) != 0 ? .on : .off }
        }
        set(ownerBoxes, (m >> 6) & 0o7)
        set(groupBoxes, (m >> 3) & 0o7)
        set(otherBoxes, m & 0o7)
        refreshOctal()
    }

    private func refreshOctal() {
        let m = currentMode()
        octalLabel.stringValue = String(format: "0%03o", m)
        octalLabel.textColor = Theme.accent
    }

    @objc private func cancel() { hide() }

    @objc private func confirm() {
        guard !paths.isEmpty else { hide(); return }
        let mode = currentMode()
        let oct = String(format: "%03o", mode)
        let recursive = recursiveBox.state == .on
        let scope = modePopup.indexOfSelectedItem   // 0 both / 1 files / 2 dirs
        let quoted = paths.map { SFTPTransfer.quote($0) }.joined(separator: " ")

        let cmd: String
        if recursive {
            // find 按类型过滤后 chmod；both 时用 chmod -R
            switch scope {
            case 1:
                cmd = "find \(quoted) -type f -exec chmod \(oct) {} + 2>&1"
            case 2:
                cmd = "find \(quoted) -type d -exec chmod \(oct) {} + 2>&1"
            default:
                cmd = "chmod -R \(oct) \(quoted) 2>&1"
            }
        } else {
            cmd = "chmod \(oct) \(quoted) 2>&1"
        }

        guard let run = execRunner else {
            onDone?("需要 SSH 会话才能改权限")
            hide()
            return
        }
        Log.info("chmod \(oct)\(recursive ? " -R" : "") \(paths.count) 项", "sftp")
        run(cmd) { [weak self] out in
            let trimmed = out.trimmingCharacters(in: .whitespacesAndNewlines)
            let msg = trimmed.isEmpty ? "已设置权限 \(String(format: "0%03o", mode))" : "chmod: \(trimmed)"
            self?.onDone?(msg)
            self?.hide()
        }
    }
}
